# SupplyMind AI — Dataset Understanding



In [1]:
# -------------------
# Imports
# -------------------

from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

from supplymind.features.predictions.domain.constants import (
    CATEGORICAL_FEATURES,
    NUMERICAL_FEATURES,
    TARGET_COLUMN,
)
from supplymind.features.predictions.ml.workflow import load_clean_syndelay

sns.set_theme(style="whitegrid")
pd.set_option("display.max_columns", 100)
pd.set_option("display.max_rows", 100)

In [2]:
# -------------------
# Project configuration
# -------------------

DATASET_PATH = Path("../data/raw/syndelay/syndelay_v1.csv")
REPORT_ROOT = Path("../reports")

assert DATASET_PATH.exists(), (
    f"Dataset not found: {DATASET_PATH}"
)

In [3]:
# -------------------
# Load canonical SynDelay
# -------------------

df = load_clean_syndelay(DATASET_PATH)
df.head(20)

,payment_type,profit_per_order,sales_per_customer,category_id,category_name,customer_city,customer_country,customer_id,customer_segment,customer_state,customer_zipcode,department_id,department_name,latitude,longitude,market,order_city,order_country,order_customer_id,order_date,order_id,order_item_cardprod_id,order_item_discount,order_item_discount_rate,order_item_id,order_item_product_price,order_item_profit_ratio,order_item_quantity,sales,order_item_total_amount,order_profit_per_order,order_region,order_state,order_status,product_card_id,product_category_id,product_name,product_price,shipping_date,shipping_mode,label,delivery_outcome,is_delayed
0,PAYMENT,-32.924488,278.95000,38,Kids' Golf Clubs,Caguas,Puerto Rico,12446.5625,Corporate,PR,725.000,6,Outdoors,18.247585,-66.370610,LATAM,Fort-de-France,Martinica,12410.3640,2015-06-19 12:00:00.000000000,62377.0100,502.0,77.9940,0.20,154588.400,129.99,-0.22,3.0,389.97,311.9760,-68.634720,Caribbean,Martinique,PENDING_PAYMENT,858,38,GolfBuddy VT3 GPS Watch,129.99,2015-06-22 12:00:00.000000000,Second Class,2,2,1
1,DEBIT,107.874500,263.98000,17,Cleats,Caguas,Puerto Rico,7782.0170,Corporate,PR,725.000,4,Apparel,18.228312,-66.370510,Africa,Luanshya,Zambia,7735.5444,2016-05-08 09:21:35.999999948,41717.2660,365.0,35.9940,0.12,102869.490,59.99,0.36,5.0,299.95,263.9560,95.024160,East Africa,Copperbelt,COMPLETE,365,17,Perfect Fitness Perfect Rip Deck,59.99,2016-05-12 09:21:35.999999948,Same Day,1,1,0
2,PAYMENT,35.770718,109.65013,17,Cleats,Caguas,Puerto Rico,7378.1113,Consumer,PR,725.000,4,Apparel,18.287943,-66.370514,Pacific Asia,Ankara,Turkey,7137.4355,2017-08-02 06:23:02.400000285,58860.4900,365.0,10.7982,0.09,150127.310,59.99,0.33,2.0,119.98,109.1818,36.029994,West Asia,Ankara,PENDING_PAYMENT,365,17,Perfect Fitness Perfect Rip Deck,59.99,2017-08-04 06:23:02.400000285,Standard Class,0,0,0
3,PAYMENT,43.587560,113.09000,18,Men's Footwear,Caguas,Puerto Rico,1448.6765,Consumer,PR,725.000,4,Apparel,18.234800,-66.370600,LATAM,Tegucigalpa,Honduras,1502.5054,2015-06-20 21:36:00.000000129,32101.9300,403.0,19.4985,0.15,82330.920,129.99,0.35,1.0,129.99,110.4915,38.672025,Central America,Francisco Morazan,PENDING_PAYMENT,403,18,Nike Men's CJ Elite 2 TD Football Cleat,129.99,2015-06-26 21:36:00.000000129,Second Class,2,2,1
4,PAYMENT,49.804802,191.98090,9,Cardio Equipment,Madison,EE. UU.,5123.5254,Corporate,WI,36384.400,3,Footwear,33.920456,-89.977650,LATAM,Leon,Nicaragua,5313.9097,2016-09-17 19:40:47.999999974,1953.9133,191.0,3.9996,0.02,4647.874,99.99,0.29,2.0,199.98,195.9804,56.834316,Central America,Leon,PENDING_PAYMENT,191,9,Nike Men's Free 5.0+ Running Shoe,99.99,2016-09-19 19:40:47.999999974,Standard Class,1,1,0
5,PAYMENT,73.840120,168.22969,48,Water Sports,Caguas,Puerto Rico,11010.9795,Consumer,PR,725.000,7,Fan Shop,18.206371,-66.370544,Pacific Asia,Sari,Iran,10985.4860,2017-05-15 21:11:31.200000138,29988.0920,1073.0,29.9985,0.15,73187.580,199.99,0.45,1.0,199.99,169.9915,76.496175,South Asia,Mazandaran,PENDING_PAYMENT,1073,48,Pelican Sunstream 100 Kayak,199.99,2017-05-17 21:11:31.200000138,Second Class,2,2,1
6,TRANSFER,95.949814,284.98000,17,Cleats,Florissant,EE. UU.,10328.7110,Consumer,MO,78467.010,4,Apparel,34.051000,-100.160970,USCA,Houston,United States,10045.0630,2015-04-03 18:43:11.999999896,35271.3900,365.0,20.9965,0.07,89811.640,59.99,0.34,5.0,299.95,278.9535,94.844190,US Center,Texas,PROCESSING,365,17,Perfect Fitness Perfect Rip Deck,59.99,2015-04-07 18:43:11.999999896,Standard Class,0,0,0
7,TRANSFER,18.990294,104.37533,9,Cardio Equipment,Pompano Beach,EE. UU.,9461.3550,Consumer,FL,60060.580,3,Footwear,29.810043,-84.439910,Pacific Asia,Wollongong,Australia,8980.4450,2017-09-27 01:26:23.999999801,41009.0230,191.0,0.0000,0.00,102109.500,99.99,0.15,1.0,99.99,99.9900,14.998500,Oceania,New South Wales,PENDING,191,9,Nike Men's Free 5.0+ Running Shoe,99.99,2017-10-03 01:26:23.999999801,First Class,2,2,1
8,CASH,59.853287,284.98000,43,Camping & Hiking,Elmhurst,EE. UU.,9096.2260,Consumer,IL,78085.6

### Finding

The raw SynDelay file has 41 columns. After canonicalization, I have 43 columns because
SupplyMind preserves the source `label` and adds `delivery_outcome` plus the binary
`is_delayed` target. The raw CSV itself is not modified.

In [4]:
# -------------------
# Dataset dimensions
# -------------------

print(f"Rows: {len(df):,}")
print(f"Columns after canonicalization: {df.shape[1]}")
print(f"Memory usage: {df.memory_usage(deep=True).sum()/1_000_000:.2f} MB")

Rows: 155,488
Columns after canonicalization: 43
Memory usage: 74.97 MB


In [5]:
# -------------------
# Data types
# -------------------

dtype_table = (
    df.dtypes.astype(str)
    .rename("dtype")
    .reset_index()
    .rename(columns={"index": "column"})
)
dtype_table

,column,dtype
0,payment_type,string
1,profit_per_order,float64
2,sales_per_customer,float64
3,category_id,int64
4,category_name,string
5,customer_city,string
6,customer_country,string
7,customer_id,float64
8,customer_segment,string
9,customer_state,string


### Finding

The canonical dataset contains a mix of numeric, categorical and two converted datetime
columns. `order_date` and `shipping_date` are now real timestamps instead of Excel serial
numbers, which makes the temporal analysis readable and lets me split the data by time.

In [6]:
# -------------------
# Full data-quality profile
# -------------------

profile = pd.DataFrame({
    "column": df.columns,
    "dtype": df.dtypes.astype(str).values,
    "missing_count": df.isna().sum().values,
    "missing_pct": (df.isna().mean().values * 100).round(3),
    "unique_count": df.nunique(dropna=True).values,
}).sort_values(
    ["missing_pct", "unique_count"],
    ascending=[False, False],
)

profile

,column,dtype,missing_count,missing_pct,unique_count
1,profit_per_order,float64,0,0.0,152935
20,order_id,float64,0,0.0,152595
7,customer_id,float64,0,0.0,148992
24,order_item_id,float64,0,0.0,132436
18,order_customer_id,float64,0,0.0,132062
13,latitude,float64,0,0.0,124885
38,shipping_date,datetime64[ns],0,0.0,118997
19,order_date,datetime64[ns],0,0.0,118809
14,longitude,float64,0,0.0,91519
2,sales_per_customer,float64,0,0.0,85233


### Finding

There are no missing values in the 155,488 rows. That is useful for this benchmark, but I
am still keeping imputers in the production preprocessing pipeline because real ERP/TMS
data will not be this clean. Several ID-like columns have extremely high cardinality, so
I will not treat them as normal categorical predictors.

In [7]:
# -------------------
# Duplicate analysis
# -------------------

duplicate_count = int(df.duplicated().sum())
duplicate_rate = duplicate_count / len(df)

pd.DataFrame({
    "duplicate_rows": [duplicate_count],
    "duplicate_rate": [duplicate_rate],
})

,duplicate_rows,duplicate_rate
0,0,0.0


### Finding

I found **0 exact duplicate rows**, so there is nothing to remove at this stage. I am not
deduplicating on customer or order IDs because repeated entities can be legitimate order
history rather than duplicate records.

In [8]:
# -------------------
# Numerical summary
# -------------------

df.describe(include="number").T

,count,mean,std,min,25%,50%,75%,max
profit_per_order,155488.0,22.625080,119.818361,-4274.980000,7.568531,32.682004,64.366681,805.540160
sales_per_customer,155488.0,181.098027,113.480905,7.780167,105.570935,165.948080,245.980000,1939.990000
category_id,155488.0,31.642847,15.597621,2.000000,18.000000,29.000000,45.000000,76.000000
customer_id,155488.0,6681.724789,4112.453329,2.000000,3297.375900,6417.069250,9735.472250,20757.000000
customer_zipcode,155488.0,35840.444698,36827.395347,0.000000,725.000000,20861.633000,77048.682500,99205.000000
department_id,155488.0,5.433641,1.627307,2.000000,4.000000,5.000000,7.000000,12.000000
latitude,155488.0,29.784330,9.695576,-33.937553,18.266945,33.362042,39.109177,48.781933
longitude,155488.0,-84.722127,20.402017,-158.025990,-97.668050,-78.744316,-66.370580,115.040000
order_customer_id,155488.0,6642.201590,4085.771174,2.000000,3272.024200,6397.487550,9672.004250,20757.000000
order_id,155488.0,36210.817670,20694.589889,1.000000,18119.014750,36280.830000,53833.608500,77204.000000


### Numerical summary

The commercial variables have wide ranges and clear outliers: for example,
`profit_per_order` reaches large negative values. I am keeping these observations rather
than trimming them automatically because losses and unusually large orders can be real
business events and tree models can handle them well.

In [9]:
# -------------------
# Categorical summary
# -------------------

df.describe(include=["object", "string"]).T

,count,unique,top,freq
payment_type,155488,4,DEBIT,62103
category_name,155488,50,Cleats,21692
customer_city,155488,563,Caguas,57992
customer_country,155488,2,EE. UU.,95426
customer_segment,155488,3,Consumer,80914
customer_state,155488,44,PR,60062
department_name,155488,11,Fan Shop,57595
market,155488,5,Europe,46156
order_city,155488,3505,New York City,2052
order_country,155488,161,United States,21653


### Categorical summary

The categorical fields vary strongly in cardinality. Low-cardinality fields such as
`payment_type`, `customer_segment`, `market` and `shipping_mode` are suitable for
one-hot encoding. City/state fields are much larger, so they need a different treatment
instead of expanding into thousands of dummy columns.

In [10]:
# -------------------
# Source outcome distribution
# -------------------

source_target = (
    df["label"]
    .value_counts()
    .sort_index()
    .rename(index={0: "Early", 1: "On-time", 2: "Delayed"})
    .to_frame("count")
)
source_target["percentage"] = (
    source_target["count"] / len(df) * 100
).round(2)

source_target

,count,percentage
label,,
Early,36724,23.62
On-time,29055,18.69
Delayed,89709,57.70


### Finding

The source outcome is clearly dominated by delayed deliveries:
**89,709 delayed (57.70%)**, **36,724 early (23.62%)** and **29,055 on-time (18.69%)**.
That makes delay the most common operational outcome in this synthetic dataset.

In [11]:
# -------------------
# Binary target distribution
# -------------------

binary_target = (
    df[TARGET_COLUMN]
    .value_counts()
    .sort_index()
    .rename(index={0: "Not delayed", 1: "Delayed"})
    .to_frame("count")
)
binary_target["percentage"] = (
    binary_target["count"] / len(df) * 100
).round(2)

binary_target

,count,percentage
is_delayed,,
Not delayed,65779,42.3
Delayed,89709,57.7


### Finding

For SupplyMind V1, early and on-time are merged into **65,779 not-delayed orders
(42.3%)**, while delayed remains **89,709 orders (57.7%)**. The split is not extreme
enough to justify SMOTE or aggressive resampling before I see how the baseline models
behave.

In [12]:
# -------------------
# Cardinality analysis
# -------------------

cardinality = (
    df.nunique(dropna=True)
    .sort_values(ascending=False)
    .rename("unique_values")
    .to_frame()
)

cardinality.head(25)

,unique_values
profit_per_order,152935
order_id,152595
customer_id,148992
order_item_id,132436
order_customer_id,132062
latitude,124885
shipping_date,118997
order_date,118809
longitude,91519
sales_per_customer,85233


### Finding

The cardinality check confirms why IDs should be excluded: `order_id`, `customer_id`
and `order_item_id` are almost unique per row. `order_city` has **3,505** values,
`order_state` **1,054**, and `customer_city` **563**, so I keep their information through
frequency encoding rather than one-hot encoding every location.

In [13]:
# -------------------
# Date coverage
# -------------------

date_coverage = pd.DataFrame({
    "column": ["order_date", "shipping_date"],
    "minimum": [df["order_date"].min(), df["shipping_date"].min()],
    "maximum": [df["order_date"].max(), df["shipping_date"].max()],
    "missing": [df["order_date"].isna().sum(), df["shipping_date"].isna().sum()],
})

date_coverage

,column,minimum,maximum,missing
0,order_date,2015-01-01 00:23:02.400000285,2018-01-31 23:36:57.599999714,0
1,shipping_date,2015-01-01 15:38:52.800000164,2018-02-06 23:36:57.599999714,0


### Finding

The dataset covers orders from **1 Jan 2015 to 31 Jan 2018**. Shipping dates extend to
**6 Feb 2018**. I use `order_date` for the chronological split because it is available at
prediction time; `shipping_date` is useful for exploration but is excluded from the model
to avoid using future information.